### ISSUE TITLE
Rule#each raise an error if the schema is extended by another one

### ISSUE THREAD

**Issue by Sbastien** [dry-rb/dry-validation#659]
Hello Guys,

First of all, thank you for your gems which are really good :)

**Describe the bug**

An error is raised when we use the `Rule#each` method for a key when another key, **defined before the array**, does not pass the schema validation :

```bash
NoMethodError: undefined method `each_with_index' for "wrong_type":String
Did you mean?  each_line
```

The error seem to be raised from here : https://github.com/dry-rb/dry-validation/blob/master/lib/dry/validation/rule.rb#L86

**To Reproduce**

```ruby
require 'dry-validation'

class TestContract < Dry::Validation::Contract
  params do
    required(:something).filled(:string)
    required(:my_array).array(:hash) do
      optional(:test).filled(:string)
    end
  end

  rule(:my_array).each do

  end
end

# `something: nil` doesnt pass the schema validation
TestContract.new.(something: nil, my_array: 'wrong_type').errors.to_h
```

**Expected behavior**

It `.each` block or the macros should not be executed when the schema is not correct.

**Your environment**

- Affects my production application: **YES**
- Ruby version: 2.7.0
- OS: MacOS

**Comment by Sbastien** [comment]
I've update the description because it was not correct.
